### DISCUSSION TITLE
Database migrations

### DISCUSSION THREAD

**Discussion by lqmanh** [supabase/supabase#134] [Questions] [answered]
- **I'm submitting a ...**
  - [ ] bug report
  - [x] feature request
  - [ ] question about the decisions made in the repository
  - [x] question about how to use this project

- **Summary**
Currently how should I handle database migrations with Supabase? Or is there any plan to support it yet?

**Comment by awalias** [comment, marked as answer]
hey @lqmanh . We plan to build git integrations which will handle migrations and two-way syncing between git and postgres

in the meantime if you go to the settings page on the dashboard you will find your postgres db details which you can use to run any existing postgres migration tools

**Reply by joshnuss** [reply]
Suggestion: Would be cool if it supported seeds as well.

And it would be cool if a supabase project could be deployed with `git push` (like heroku).

Imagining a workflow like:

```
> gh repo fork supabase/some-example-project
> git remote add supabase https://git.supabase.com/my-project name
> git push supabase main
Migrating database...
Loading seeds...
Done.
```

See https://github.com/joshnuss/git-ssh-server for a rough example of a git server than can run code when receiving a push.

**Comment by kachar** [comment]
I find this supabase article relevant to the discussion https://supabase.io/blog/2021/03/31/supabase-cli#migrations

### Schema migrations via diff

* https://github.com/supabase/pgadmin4/blob/cli/web/cli.py
* https://hub.docker.com/r/supabase/pgadmin-schema-diff

```
docker run supabase/pgadmin-schema-diff \
  'postgres://user:pass@local:5432/diff_source' \
  'postgres://user:pass@production:5432/diff_target' \
  > diff_demo.sql

Starting schema diff...
Comparision started......0%
Comparing Event Triggers...2%
Comparing Extensions...4%
Comparing Languages...8%
Comparing Foreign Servers...14%
Comparing Foreign Tables of schema 'public'...28%
Comparing Tables of schema 'public'...50%
Comparing Domains of schema 'test_schema_diff'...66%
Comparing Foreign Tables of schema 'test_schema_diff'...68%
Comparing FTS Templates of schema 'test_schema_diff'...76%
Comparing Functions of schema 'test_schema_diff'...78%
Comparing Procedures of schema 'test_schema_diff'...80%
Comparing Tables of schema 'test_schema_diff'...90%
Comparing Types of schema 'test_schema_diff'...92%
Comparing Materialized Views of schema 'test_schema_diff'...96%
Done.

## the diff is written to diff_demo.sql
```

**Reply by soedirgo** [reply]
Yup, we're on track on supporting this in the CLI in the next few months. Stay tuned :)

**Reply by mstade** [reply]
Hi @soedirgo!

We're trying to run the aforementioned command, but it seems to fail to connect to our databases due to us having passwords with special characters (specifically: `@` seems to cause confusions.) These passwords were obviously generated  at the time of creating the databases, and we can't seem to change them from the dashboard now to remove special characters.

What's our recourse, here? Is it just a matter of changing the password of the `postgres` role via SQL? Is there any particular reason why this password can't be changed via the dashboard?

**Reply by mstade** [reply]
For the record: we've resorted to using the pgAdmin schema diff UI for now, but we'd like to do this in the terminal in the future, so we're not SOL or anything right now.

**Reply by soedirgo** [reply]
Hmm, does it work if you percent-encode the password (i.e. change `@` to `%40`)? Alternatively, I suppose you can `ALTER ROLE postgres WITH PASSWORD 'foo'` in the SQL editor.

**Comment by baptisteArno** [comment]
The problem with schema migrations via diff is that it won't migrate the data content, right?

Let's say I have a schema with an attribute `url` which is a string, let's say I want to migrate this to `urls` which would be an array of string. We need the existing `url` to be converted into an array of one item

What would be the best practice here with Supabase?

**Reply by soedirgo** [reply]
Yes, we don't handle data migration with the diff. What you *can* do however is to edit the migration generated by the diff before deploying it, that way you can do whatever you need to do (rename a column, backfill old data, etc.).

All of this is just vaporware though, it's still WIP in [here](https://github.com/supabase/cli/tree/new).

**Reply by TomasHubelbauer** [reply]
Up to date link: https://github.com/supabase/cli/tree/main/examples/tour

**Reply by 0-vortex** [reply]
thanks for this!

**Reply by MentalGear** [reply]
Hey @soedirgo are data migrations/transforms as described by @baptisteArno now supported on supabase?
I couldn't find anything in the docs on this specifically and the above link 404s.

**Comment by idosius** [comment]
I guess migrations are supported now?

https://supabase.com/docs/reference/cli/usage#supabase-migration

https://www.youtube.com/watch?v=Kx5nHBmIxyQ

**Reply by soedirgo** [reply]
Yup, we now support managing migrations.

**Reply by josephbuchma** [reply]
What about "down" migrations? I couldn't find any info in docs

**Reply by soedirgo** [reply]
It's forward-only, so you need another migration to revert an earlier migration.

**Reply by lsarni** [reply]
@soedirgo `supabase migration repair <version> --status reverted` is this command in any way related to a "down" migration? If not, could you link me to an explanation? I failed to found doc about it

**Reply by soedirgo** [reply]
It deletes the row associated with the migration in the migration history `supabase_migrations.schema_migrations` - so you *could* do a manual down migration by manually reverting the migration directly with SQL, and then running `supabase migration repair`. Then the CLI will behave as if the migration hasn't been executed.

**Comment by kinthaiofficial** [comment]
Database migrations with Supabase in a team workflow has a few rough edges worth thinking through:

**The core tension:**
Supabase's Studio UI lets you make schema changes directly — convenient for solo devs, problematic for teams where multiple people are making changes that need to be coordinated.

**The pattern that works:**
Treat the Studio as read-only for production. All schema changes go through migration files. Supabase CLI's `supabase db diff` generates the migration file from what you've done in a local Studio instance, which you then commit.

```bash
# After making changes in local Studio:
supabase db diff --use-migra -f my_migration_name > supabase/migrations/$(date +%Y%m%d%H%M%S)_my_migration_name.sql
# Review, commit, and apply via CI
```

**Team coordination**:
- One person owns the migration sequence. Review migrations in PRs like you review code.
- Linear migration numbering (timestamp-prefixed) prevents the "two developers created migration 005 on the same branch" problem.
- Squash migrations before merging long-lived feature branches to keep history clean.

**Remote vs local schemas diverging:**
When someone makes a change directly in production Studio (it happens), run `supabase db diff` between production and your local schema to generate a "sync migration" that brings your repo back in sync.

**For AI agents writing to Supabase:**
Schema flexibility matters more — consider using a staging schema (`agent_staging`) where agents can create temporary tables without touching the production schema. RLS policies restrict what agents can access.

What's the team size and branching model? That affects which migration strategy makes sense.
